File: GuardianMonsters/guardians/src/de/limbusdev/guardianmonsters/guardians/monsters/AGuardianFactory.java
package de.limbusdev.guardianmonsters.guardians.monsters;

import java.util.ArrayList;
import java.util.Collection;
import java.util.UUID;

/**
 * AGuardianFactory
 *
 * Design Pattern: Factory Method
 *
 * @author Georg Eckert 2017
 */

public abstract class AGuardianFactory
{
    private static Collection<String> UUIDs;

    public AGuardianFactory()
    {
        getUUIDs();
    }

    // ............................................................................................. IMPLEMENTED METHODS

    /**
     * Creates a unique identifier for a new Guardian
     * @return
     */
    public final static String createNewUUID()
    {
        boolean unique = false;
        String newUUID = "";
        while(!unique)
        {
            newUUID = UUID.randomUUID().toString();
            if(!getUUIDs().contains(newUUID))
            {
                unique = true;
            }
        }
        getUUIDs().add(newUUID);
        return newUUID;
    }


    public final static Collection<String> getUUIDs()
    {
        if(UUIDs == null) UUIDs = new ArrayList<>();
        return UUIDs;
    }


    // ............................................................................................. ABSTRACT METHODS

    /**
     * FACTORY METHOD
     *
     * Takes full responsibility of creating a guardian.
     *
     * @param ID    ID of the guardian to be created, defines species
     * @param level level if the guardian to be created
     * @return      the complete generated guardian
     */
    public abstract AGuardian createGuardian(int ID, int level);    // P04: Factory Method
}

File: GuardianMonsters/guardians/src/de/limbusdev/guardianmonsters/guardians/monsters/AGuardianFactory.kt
package de.limbusdev.guardianmonsters.guardians.monsters

import de.limbusdev.guardianmonsters.guardians.GuardiansServiceLocator
import java.util.*

/**
 * AGuardianFactory
 *
 * Design Pattern: Factory Method
 *
 * @author Georg Eckert 2017
 */

abstract class AGuardianFactory protected constructor()
    : GuardiansServiceLocator.Service
{
    // ............................................................................................. ABSTRACT METHODS

    /**
     * FACTORY METHOD
     *
     * Takes full responsibility of creating a guardian.
     *
     * @param ID    ID of the guardian to be created, defines species
     * @param level level if the guardian to be created
     * @return      the complete generated guardian
     */
    abstract fun createGuardian(ID: Int, level: Int): AGuardian     // P04: Factory Method

    override fun destroy() {UUIDs.removeAll(UUIDs)}

    companion object
    {
        val UUIDs: MutableCollection<String> = ArrayList()

        // ......................................................................................... IMPLEMENTED METHODS

        /**
         * Creates a unique identifier for a new Guardian
         * @return
         */
        fun createNewUUID(): String
        {
            var unique = false
            var newUUID = ""
            while(!unique)
            {
                newUUID = UUID.randomUUID().toString()
                if(!UUIDs.contains(newUUID))
                {
                    unique = true
                }
            }
            UUIDs.add(newUUID)
            return newUUID
        }
    }
}
